include/envoy/stream_info/stream_info.h

```diff
@@ -481,6 +481,16 @@ class StreamInfo {
    */
   virtual FilterState& filterState() PURE;
   virtual const FilterState& filterState() const PURE;
+  virtual std::shared_ptr<FilterState> filterStatePtr() PURE;
```

Comment:
Can you just make the reference method above return the shared_ptr? You can return a const reference to the shared_ptr which should make this just as efficient in the common case. Then we don't need this new method.

Reply:
@mattklein123 : Won't that mean changing a lot more files because we would be changing a commonly used signature?
 I agree it would be better going forward...

Reply:
I doubt there are that many call-sites. I would just fix them.

Reply:
Okay .. will prepare a pr with that

Reply:
define FilterStateSharedPtr (i.e. `using FilterStateSharedPtr = std::shared_ptr<FilterState>;`) in filter_state.h and use it everywhere possible.

Reply:
@lizan : will do.
@mattklein123 : Am going with following two signatures:

using FilterStateSharedPtr = std::shared_ptr<FilterState>;) 
FilterStateSharedPtr filterState() override { return filter_state_; } 
const FilterStateSharedPtr filterState() const override { return filter_state_;}

Returning a ref to shared ptr.. I am getting following error:
bazel-out/k8-fastbuild/bin/source/common/stream_info/_virtual_includes/stream_info_lib/common/stream_info/stream_info_impl.h:220:77: error: returning reference to local temporary object [-Werror,-Wreturn-stack-address]
  const std::shared_ptr<FilterState>& filterState() const override { return filter_state_;}
 which looks like can happen in c++14 (https://stackoverflow.com/a/47167649)
also, it's better to return shared_ptr by value as compared to reference, so as to avoid lifecycle issues (https://stackoverflow.com/questions/10643563/how-to-return-smart-pointers-shared-ptr-by-reference-or-by-value)

please let me know what you think?

Reply:
`const FilterStateSharedPtr filterState() const override { return filter_state_;}` is not correct, as you are returning a const shared_ptr by value. Just leave the const version as it was, or change it to return a shared_ptr to a const object.

> also, it's better to return shared_ptr by value as compared to reference, so as to avoid lifecycle issues

We do this in many places to avoid the copy/inc where necessary. Please return by const ref. Thanks!

Reply:
Thanks for the feedback. PTAL at the PR again!

(I have to say, I learnt about constness with respect to shared_ptr's)